content/blog/how-to-build-globally-distributed-applications-with-azure-cosmos-db-and-pulumi/index.md

```diff
@@ -0,0 +1,333 @@
+            username: registry.adminUsername,
+            password: registry.adminPassword,
+        },
+    });
+
+    return ({ location }: RegionalContext) => {
+        // ... see the next sample
+    };
+}
+```
+
+Now, the factory function that builds per-region infrastructure can directly reference the Registry and the image to deploy an Azure Container Instances group. Note how registry and dockerImage are used as any other TypeScript variable inside the closure factory function:
+
+```ts
+function buildShoppingCartApp({ cosmosAccount, database, container }: GlobalContext) {
+
+    const registry = new azure.containerservice.Registry(/* ... */);
+    const dockerImage = new docker.Image(/* ... */);
+
+    return ({ location }: RegionalContext) => {
+        const group = new azure.containerservice.Group(`aci-${location}`, {
+            resourceGroupName: resourceGroup.name,
+            location,
+            imageRegistryCredentials: [{
+                server: registry.loginServer,
+                username: registry.adminUsername,
+                password: registry.adminPassword,
+            }],
+            osType: "Linux",
+            containers: [
+                {
+                    cpu: 0.5,
+                    image: dockerImage.imageName,
+                    memory: 1.5,
+                    name: "web-server",
+                    ports: [{
+                        port: 80,
+                        protocol: "TCP",
+                    }],
+                    environmentVariables: {
+                        ENDPOINT: cosmosAccount.endpoint,
+                        MASTER_KEY: cosmosAccount.primaryMasterKey,
+                        DATABASE: database.name,
+                        COLLECTION: container.name,
+                        LOCATION: location,
+                    },
+                },
+            ],
+            ipAddressType: "public",
+            dnsNameLabel: `acishop-${location}`,
+        });
+
+        return {
+            url: group.fqdn,
+        };
+    };
+}
+```
+
+Finally, a `CosmosApp` is defined:
+
+```ts
+const aci = new CosmosApp("aci", {
+    resourceGroup,
+    locations,
+    databaseName: "cartdb",
+    containerName: "items",
+    factory: buildShoppingCartApp,
+    enableMultiMaster: true,
+});
+```
+
+Shopping Cart workload expects both reads and writes from the end-users, so the team decides to enable multi-master support in their Cosmos DB instance.
+
+[ [Full Example](https://github.com/pulumi/examples/tree/master/azure-ts-cosmosapp-component/aci.ts) ]
+
+## Cosmos App with Virtual Machine Scale Sets
+
+The Pricing Engine team has to go through the same steps but employ IaaS-based infrastructure to run application code. The factory function creates a dozen of resources related to virtual machines, load balancing, networking, and auto-scaling. You may define as many resources as needed in the code, as long as it returns a pointer to the proper resource to link the endpoint.
```

Comment:
```suggestion
The Pricing Engine team has to go through the same steps but employ IaaS-based infrastructure to run application code. The factory function creates a dozen resources related to virtual machines, load balancing, networking, and auto-scaling. You may define as many resources as needed in the code, as long as it returns a pointer to the proper resource to link the endpoint.
```

Reply:
My dictionary suggests "of" here, as some examples that I see online. Is there a rule about that?

Reply:
It's generally either "dozens of donuts" or "a dozen donuts", but not "a dozen of donuts." It's fine, however, to say, "A bag of donuts".  

Honestly, I have no idea why, and I'd love to have a better explanation than that. Maybe it has something to do with "dozen" being more of a counting word, like "twelve" (so like "twelve donuts" rather than "twelve of donuts". But then, "twelve of _those_ donuts" is okay?  

Ugh. English 🙄 .